### PR TITLE
ESC Setup / programming (bidirectional telemetry / forward programming) - implemented for HW5 ESCs

### PR DESCRIPTION
**Pull request by bob01**
ESC setup / programming (forward programming) using the Tx via bidirectional telemetry #101

Complete for HobbyWing V5

- Implemented on common request / response finite state machine (rrfsm), uses common ESC parameter read and write pattern and common MSP transport as #108 

### Configuration
- same as #108 

### Code
- TELEMETRY_BUFFER_SIZE increased to accommodate larger HW5 responses
- rrfsm protocol callback signature changed to accept time in Us vs Ms for protocols like HW5 which need to see the higher resolution timer
- out of band metadata value sent every 128 tele frame as ESC status value  that Tx widgets can use to...
   - identify ESC (for supported ESCs)
   - signal that ESC needs to be reset (for new settings to take effect / unsolicited telemetry to resume)
   - or can be ignored.

#### HW5 support
- configure FC as 'HobbyWing v5'
- requires Platinum 5 ESC, tested with PL-04.0.11 and current version PL-04.1.02
- ESC can be configured at any time, restart required - ESC must be power-cycled e.g.
   - bench setup or adjustments in the pits
 
image
image
image
image
image
image
image
image

### CODE REVIEW

File: src/main/sensors/esc_sensor.c [hunk 1]
```diff
@@ -1986,8 +2218,9 @@ static bool tribDecode(timeMs_t currentTimeMs)
     }
 }
 
-static bool tribCrankUncSetup(timeMs_t currentTimeMs)
+static bool tribCrankUncSetup(timeMs_t currentTimeUs)
```
Comment (rotorflight) [inline src/main/sensors/esc_sensor.c:2221]: Should be `timeUs_t`

File: src/main/sensors/esc_sensor.c [hunk 2]
```diff
@@ -1448,9 +1303,9 @@ static void apdSensorProcess(timeUs_t currentTimeUs)
 //   0: continue accepting
 typedef int8_t (*rrfsmAcceptCallbackPtr)(uint16_t c);
 
-typedef bool (*rrfsmStartCallbackPtr)(timeMs_t currentTimeMs);  // return true to continue w/ default initialization (if in doubt return true)
-typedef bool (*rrfsmDecodeCallbackPtr)(timeMs_t currentTimeMs); // return true if frame was decoded successfully
-typedef bool (*rrfsmCrankCallbackPtr)(timeMs_t currentTimeMs);  // return true to continue w/ default loop (advanced, if in doubt return true)
+typedef bool (*rrfsmStartCallbackPtr)(timeMs_t currentTimeUs);  // return true to continue w/ default initialization (if in doubt return true)
```
Comment (rotorflight) [inline src/main/sensors/esc_sensor.c:1306]: Should be `timeUs_t`

File: src/main/sensors/esc_sensor.c [hunk 3]
```diff
@@ -1448,9 +1303,9 @@ static void apdSensorProcess(timeUs_t currentTimeUs)
...
+typedef bool (*rrfsmDecodeCallbackPtr)(timeUs_t currentTimeUs); // return true if frame was decoded successfully
+typedef bool (*rrfsmCrankCallbackPtr)(timeMs_t currentTimeUs);  // return true to continue w/ default loop (advanced, if in doubt return true)
```
Comment (rotorflight) [inline src/main/sensors/esc_sensor.c:1308]: `timeUs_t`

File: src/main/sensors/esc_sensor.c [hunk 4]
```diff
@@ -1583,6 +1438,380 @@ static void rrfsmSensorProcess(timeUs_t currentTimeUs)
...
+    reqLength = len;
+
+    rrfsmFramePeriod = framePeriod;
+    rrfsmFrameTimeout = frameTimeout;
+
+    return true;
+}
+
+static void pl5BuildNextReq()
+{
```
Comment (rotorflight) [inline src/main/sensors/esc_sensor.c:1599]: `(void)`

**Review by rotorflight — CHANGES_REQUESTED** [review]
Minor tweaks needed

File: src/main/sensors/esc_sensor.c (hunk 4 above)
Comment (bob01) [reply src/main/sensors/esc_sensor.c:1599]: timeUs_t and fn(void) usage addressed

**Review by rotorflight — APPROVED** [review]

**Pull request merged**